types/datum.go

```diff
@@ -848,7 +848,17 @@ func ProduceStrWithSpecifiedTp(s string, tp *FieldType, sc *stmtctx.StatementCon
 					}
 					runeCount++
 				}
-				err = ErrDataTooLong.GenWithStack("Data Too Long, field len %d, data len %d", flen, characterLen)
+
+				truncateTrailingSpacesLen := math.MaxInt32
+				if tp.Tp == mysql.TypeString {
```

Comment:
we should also consider binary strings, consider this case:
```sql
drop table if exists t;
create table t(a char(2), b char(2) binary, c binary(2));
insert into t values("aa ", null, null);
insert into t values(null, "aa ", null);
insert into t values(null, null, "aa ");
```

In MySQL, only the last insert statement fails:
```sql
MySQL(root@localhost:test) > drop table if exists t;
Query OK, 0 rows affected (0.08 sec)

MySQL(root@localhost:test) > create table t(a char(2), b char(2) binary, c binary(2));
Query OK, 0 rows affected (0.11 sec)

MySQL(root@localhost:test) > insert into t values("aa ", null, null);
Query OK, 1 row affected (0.02 sec)

MySQL(root@localhost:test) > insert into t values(null, "aa ", null);
Query OK, 1 row affected (0.04 sec)

MySQL(root@localhost:test) > insert into t values(null, null, "aa ");
ERROR 1406 (22001): Data too long for column 'c' at row 1
MySQL(root@localhost:test) > select version();
+-----------+
| version() |
+-----------+
| 8.0.11    |
+-----------+
1 row in set (0.00 sec)
```

Reply:
I tested the above statements and the results are the same.

```
mysql> create table t(a char(2), b char(2) binary, c binary(2));
Query OK, 0 rows affected (0.54 sec)

mysql> insert into t values("aa ", null, null);
Query OK, 1 row affected (0.04 sec)

mysql> insert into t values(null, "aa ", null);
Query OK, 1 row affected (0.01 sec)

mysql> insert into t values(null, null, "aa ");
ERROR 1406 (22001): Data too long for column 'c' at row 1
mysql> select version();
+------------------+
| version()        |
+------------------+
| 5.7.25-TiDB-None |
+------------------+
1 row in set (0.00 sec)

```

Reply:
@zz-jason  Is it Right?